examples/graph/gnn_citations.py

```diff
@@ -0,0 +1,728 @@
+"""
+
+# Create an edges array (sparse adjacency matrix) of shape [2, num_edges].
+edges = citations[["source", "target"]].to_numpy().T
+# Create an edge weights array of ones.
+edge_weights = tf.ones(shape=edges.shape[1])
+# Create a node features array of shape [num_nodes, num_features].
+papers = papers.sort_values("paper_id")
+node_features = tf.cast(papers[feature_names].to_numpy(), dtype=tf.dtypes.float32)
+# Create graph info tuple with node_features, edges, and edge_weights.
+graph_info = (node_features, edges, edge_weights)
+
+print("Edges shape:", edges.shape)
+print("Number of nodes:", node_features.shape[0])
+print("Number of features per node:", node_features.shape[1])
+
+"""
+### Implement a graph convolution layer
+
+We implement a graph convolution module as a [Keras Layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer?version=nightly).
+Our `GraphConvLayer` performs the following steps:
+
+1. **Prepare**: The input node representations are processed using a FFN. You can simplify
+the processing by only applying linear transformation to the representations.
+2. **Aggregate**: The node representations of each neighbours of each node are aggregated with
+respect to the `edge_weigts` using a *permutation invariant* pooling operation, such as *sum*, *mean*, and *max*,
+to prepare a single message for each node.
+3. **Update**: The `node_repesentations` and `aggregated_messages`—both of shape `[num_nodes, representation_dim]`—
+are combined and processed to produce the new state of the node representations (node embeddings).
+If `combination_type` is `gru`, the `node_repesentations` and `aggregated_messages` are stacked to create a sequence,
+then processed by a GRU layer. Otherwise, the `node_repesentations` and `aggregated_messages` are added
+or concatenated, then processed using a FFN.
+
+
+The technique implemented use ideas from [Graph Convolutional Networks](https://arxiv.org/abs/1609.02907),
+[GraphSage](https://arxiv.org/abs/1706.02216), [Graph Isomorphism Network](https://arxiv.org/abs/1810.00826),
+[Simple Graph Networks](https://arxiv.org/abs/1902.07153), and
+[Gated Graph Sequence Neural Networks](https://arxiv.org/abs/1511.05493).
+Two other key techniques that are not covered are [Graph Attention Networks](https://arxiv.org/abs/1710.10903)
+and [Message Passing Neural Networks](https://arxiv.org/abs/1704.01212).
+"""
+
+
+class GraphConvLayer(keras.layers.Layer):
+    def __init__(
+        self,
+        hidden_units,
+        dropout_rate=0.2,
+        aggregation_type="mean",
+        combination_type="concat",
+        normalize=False,
+        *args,
+        **kwargs,
+    ):
+        super(GraphConvLayer, self).__init__(*args, **kwargs)
+
+        self.hidden_units = hidden_units
+        self.aggregation_type = aggregation_type
+        self.combination_type = combination_type
+        self.dropout_rate = dropout_rate
+        self.normalize = normalize
+
+    def build(self, input_shape):
```

Comment:
Same comment about `build` -- you can do this in the constructor since you're just creating unbuilt layers and you're not using `input_shape`

Reply:
Done